### ISSUE TITLE
热词功能后容易缺少字符

### ISSUE THREAD

**Issue by jqhe-Autel** [pengzhendong/streaming-sensevoice#42]
hello，感谢大佬分享
我把您这边实现的beam-search相关的内容添加到funasr的模块里面去，但是测试的时候发现增加了热词功能后同样的中文测试集cer反而高了，然后我对比了下，发现增加hotword后，hotword后面的字符容易丢失，咨询下这个问题您有什么观点吗？比如是算法原理的问题呢 还是可能会存在一些bug呢

**Comment by pengzhendong** [comment]
可以发一条 badcase 我试试看

**Comment by jqhe-Autel** [comment]
> 可以发一条 badcase 我试试看
GT：先标记一下现在的位置然后拍照
这个是没有添加热词的结果：
HYP: 先标记一下现在的位置然后拍*
这个是添加热词”标记“后的结果：
HYP: 先标记*****位置然后拍*
有些语句不添加热词，指使用beam-search就会少字符

**Comment by jqhe-Autel** [comment]
> 可以发一条 badcase 我试试看

GT：先标记一下现在的位置然后拍照
这个是没有添加热词的结果：
HYP: 先标记一下现在的位置然后拍*
这个是添加热词”标记“后的结果：
HYP: 先标记*****位置然后拍*
有些语句不添加热词，指使用beam-search就会少字符  @pengzhendong

**Comment by pengzhendong** [comment]
请问是如何集成到 funasr 中的呢

**Comment by jqhe-Autel** [comment]
> 请问是如何集成到 funasr 中的呢

上面个例子是用streaming-sensevoice跑的，整段音频送入，没做vad和chunk拆分

集成的方式funasr的sensevoice的decoder方式是
`yseq = x.argmax(dim=-1)
yseq = torch.unique_consecutive(yseq, dim=-1)`
替换成
` token_int = self.decoder.ctc_prefix_beam_search(
                    x, beam_size=self.beam_size, is_last=True
                )`
@pengzhendong

**Comment by pengzhendong** [comment]
麻烦上传下测试音频

**Comment by jqhe-Autel** [comment]
[80.zip](https://github.com/user-attachments/files/19928341/80.zip)
@pengzhendong

**Comment by sandro-qiang** [comment]
这个原因是`asr_decoder.CTCDecoder`在匹配完一个热词后没有做`context_graph.finalize`，仅仅在`is_last`的时候做了。

我现在的临时处理是subclass了下`CTCDecoder`

```
class CTCDecoderEx(CTCDecoder):
      def __init__(
          self,
          contexts: List[str] = None,
          symbol_table: Dict[str, int] = None,
          bpe_model: str = None,
          context_score: float = 6.0,
          blank_id: int = 0,
      ):
          self.context_graph = None
          if contexts is not None:
              self.context_graph = ContextGraph(contexts, symbol_table, bpe_model)
          self.context_token_ids: List[List[int]] = tokenize(
              contexts, symbol_table, bpe_model
          )
          self.blank_id = blank_id
          self.cur_t = 0
          self.cur_hyps = []
          self.reset()

      def ctc_prefix_beam_search(
          self, ctc_probs: torch.tensor, beam_size: int, is_last: bool = False
      ):
          res = super().ctc_prefix_beam_search(
              ctc_probs, beam_size=beam_size, is_last=is_last
          )

          if not is_last and self.context_graph is not None:
              for i, hyp in enumerate(self.cur_hyps):
                  for j, token_ids in enumerate(self.context_token_ids):
                      hyp_tokens = list(hyp[0])
                      if (
                          len(hyp_tokens) >= len(token_ids)
                          and hyp_tokens[-len(token_ids) :] == token_ids
                      ):
                          score, new_state = self.context_graph.finalize(
                              hyp[1].context_state
                          )
                          hyp[1].context_score = score
                          hyp[1].context_state = new_state
                          break

          return res
```

这样处理后就不会吞字了。

**Comment by sandro-qiang** [comment]
> 这个原因是`asr_decoder.CTCDecoder`在匹配完一个热词后没有做`context_graph.finalize`，仅仅在`is_last`的时候做了。
> 
> 我现在的临时处理是subclass了下`CTCDecoder`
> 
> ```
> class CTCDecoderEx(CTCDecoder):
>       def __init__(
>           self,
>           contexts: List[str] = None,
>           symbol_table: Dict[str, int] = None,
>           bpe_model: str = None,
>           context_score: float = 6.0,
>           blank_id: int = 0,
>       ):
>           self.context_graph = None
>           if contexts is not None:
>               self.context_graph = ContextGraph(contexts, symbol_table, bpe_model)
>           self.context_token_ids: List[List[int]] = tokenize(
>               contexts, symbol_table, bpe_model
>           )
>           self.blank_id = blank_id
>           self.cur_t = 0
>           self.cur_hyps = []
>           self.reset()
> 
>       def ctc_prefix_beam_search(
>           self, ctc_probs: torch.tensor, beam_size: int, is_last: bool = False
>       ):
>           res = super().ctc_prefix_beam_search(
>               ctc_probs, beam_size=beam_size, is_last=is_last
>           )
> 
>           if not is_last and self.context_graph is not None:
>               for i, hyp in enumerate(self.cur_hyps):
>                   for j, token_ids in enumerate(self.context_token_ids):
>                       hyp_tokens = list(hyp[0])
>                       if (
>                           len(hyp_tokens) >= len(token_ids)
>                           and hyp_tokens[-len(token_ids) :] == token_ids
>                       ):
>                           score, new_state = self.context_graph.finalize(
>                               hyp[1].context_state
>                           )
>                           hyp[1].context_score = score
>                           hyp[1].context_state = new_state
>                           break
> 
>           return res
> ```
> 
> 这样处理后就不会吞字了。

哦，上面的处理是错的，原因是`ContextGraph`的`context_score`参数，这个参数越大，就越容易吞字，或者出现重复的字。不过越小就越容易短的热词不起作用。

**Comment by jqhe-Autel** [comment]
> > 这个原因是`asr_decoder.CTCDecoder`在匹配完一个热词后没有做`context_graph.finalize`，仅仅在`is_last`的时候做了。
> > 我现在的临时处理是subclass了下`CTCDecoder`
> > ```
> > class CTCDecoderEx(CTCDecoder):
> >       def __init__(
> >           self,
> >           contexts: List[str] = None,
> >           symbol_table: Dict[str, int] = None,
> >           bpe_model: str = None,
> >           context_score: float = 6.0,
> >           blank_id: int = 0,
> >       ):
> >           self.context_graph = None
> >           if contexts is not None:
> >               self.context_graph = ContextGraph(contexts, symbol_table, bpe_model)
> >           self.context_token_ids: List[List[int]] = tokenize(
> >               contexts, symbol_table, bpe_model
> >           )
> >           self.blank_id = blank_id
> >           self.cur_t = 0
> >           self.cur_hyps = []
> >           self.reset()
> > 
> >       def ctc_prefix_beam_search(
> >           self, ctc_probs: torch.tensor, beam_size: int, is_last: bool = False
> >       ):
> >           res = super().ctc_prefix_beam_search(
> >               ctc_probs, beam_size=beam_size, is_last=is_last
> >           )
> > 
> >           if not is_last and self.context_graph is not None:
> >               for i, hyp in enumerate(self.cur_hyps):
> >                   for j, token_ids in enumerate(self.context_token_ids):
> >                       hyp_tokens = list(hyp[0])
> >                       if (
> >                           len(hyp_tokens) >= len(token_ids)
> >                           and hyp_tokens[-len(token_ids) :] == token_ids
> >                       ):
> >                           score, new_state = self.context_graph.finalize(
> >                               hyp[1].context_state
> >                           )
> >                           hyp[1].context_score = score
> >                           hyp[1].context_state = new_state
> >                           break
> > 
> >           return res
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 这样处理后就不会吞字了。
> 
> 哦，上面的处理是错的，原因是`ContextGraph`的`context_score`参数，这个参数越大，就越容易吞字，或者出现重复的字。不过越小就越容易短的热词不起作用。

有什么建议的处理方式吗？ 避免热词本身起的作用不明显，还造成了其他字符的丢失

**Comment by pengzhendong** [comment]
根因找到了：`asr_decoder` 的 `CTCDecoder.backoff_context()` 只在 `is_last=True` 时调用，流式场景下，热词完全匹配后 context state 停在 end node，下一个 chunk 的新 token 如果不匹配热词的后续，`forward_one_step` 沿 fail 弧回退，score = 新node.node_score - 旧node.node_score 会变成负数，导致热词后面的字符被扣分、在 beam search 中被丢弃。

修复已推送到 asr-decoder master：
https://github.com/pengzhendong/asr-decoder/commit/a18f50e

两个改动：
1. `backoff_context()` 从 if is_last 里提出来，每次 beam search 结束都调
2. `backoff_context()` 只 finalize 已完全匹配的热词（is_end=True），保留部分匹配的状态（支持跨 chunk 匹配热词）

更新 asr-decoder 后即可生效：
```
pip install -U asr-decoder
```

测试你的 `80.wav`（GT: "先标记一下现在的位置然后拍照"）：
- 无热词: 先标记一下现在的位置然后拍照
- 热词"标记": 先标记一下现在的位置然后拍照
- 热词"拍照": 先标记一下现在的位置然后拍照

均正确，不再丢字。

**Comment by pengzhendong** [comment]
fixed in https://github.com/pengzhendong/asr-decoder/commit/a18f50e
